File: MATLAB pipeline/codes_for_analysis/new_processing_pipeline/parse_ripple.py
import math
import os
import datetime
import re
import numpy as np
import psutil
import platform
import scipy.io
import shutil


# add python version of neuroshare to path
import sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../..', 'neuroshare', 'python')))


from ns_get_analog_data_block import ns_GetAnalogDataBlock
from ns_get_analog_info import ns_GetAnalogInfo
from ns_open_file import *
import time

sys.path.append(os.path.join(os.path.dirname(__file__)))

from utils import MCW_location, get_bundle
from reylab_custompath import reylab_custompath

def parse_ripple(filenames, hours_offset=5, remove_chs=None, macro=None, 
                 max_memo_GB=None, overwrite=False, 
                 channels=None, which_system_micro='RIP'):
    # This code requires the neuroshape library in the path.
    # max_memo_GB is an idea of the number of GB allocated for the data to be
    # stored in RAM, so it is used to compute the number of segments in which
    # the data should be split for processing

    if which_system_micro is None or which_system_micro == '':
        which_system_micro = 'RIP'

    name = platform.uname()[1]
    if os.name == 'nt':
        current_user = os.environ['USERNAME']
    elif os.name == 'posix':
        current_user = os.environ['USER']

    dir_base = f'/home/{current_user}/Documents/GitHub/codes_emu'
    if os.path.exists(dir_base):
        sys.path.append(dir_base)
    custompath = reylab_custompath(['tasks/locations/'])

    if 'REYLAB' in name:
        params = MCW_location(['MCW-' + which_system_micro])
        hours_offset = params['offset']

    if hours_offset is None:
        hours_offset = 0

    if filenames is None:
        aux = os.listdir('.')
        filenames = [f for f in aux if f.endswith('.ns5')]

    expr2remove = '-\d+$'

    if macro is None:
        macro = []

    if remove_chs is None:
        remove_chs = []

    if channels is None:
        channels = []

    if os.name == 'nt':
        memo_available = int(np.floor(psutil.virtual_memory().available * 0.80))
    elif os.name == 'posix':
        memo_available = 12 * (1024) ** 3  # reduce if necessary
    else:
        memo_available = 12 * (1024) ** 3  # reduce if necessary

    if max_memo_GB is not None and max_memo_GB != '':
        max_memo = max_memo_GB * (1024) ** 3
        if max_memo > memo_available:
            raise ValueError('max_memo_GB > 80% of Physical Memory Available')
    else:
        max_memo = memo_available

    t_begin = time.time()

    if isinstance(filenames, str):
        filenames = [filenames]
    folder_path = os.path.dirname(filenames[0])
    metadata_file = os.path.join(folder_path, 'NSx.mat')
    if os.path.exists(metadata_file):
        metadata = scipy.io.loadmat(metadata_file, squeeze_me=True)
    else:
        metadata = []

    new_files = []
    for fi, filename in enumerate(filenames):
        new_files.append({'name': filename})
        if len(filename) < 3 or (not filename[1:3] == ':\\' and not filename[0] == '/' 
            and not filename[1] == '/' and not filename[0:2] == '\\\\' and not filename[1:3] == ':/'):
            filename = os.path.join(os.getcwd(), filename)

        ns_status, hFile = ns_OpenFile(filename, 'single')

        with open(filename, 'rb') as fid:
            fid.seek(294, os.SEEK_SET)
            Date = np.fromfile(fid, dtype=np.uint16, count=8)

        Start_Time = datetime.datetime(Date[0], Date[1], Date[3], Date[4], Date[5], Date[6]) - datetime.timedelta(hours=hours_offset)
        Rec_length_sec = hFile['TimeSpan'] / 30000
        End_Time = Start_Time + datetime.timedelta(seconds=Rec_length_sec)

        Date_Time = [Start_Time.strftime("%Y-%m-%d %H:%M:%S"), End_Time.strftime("%Y-%m-%d %H:%M:%S")]

        if ns_status == 'ns_FILEERROR':
            raise ValueError(f'Unable to open file: {filename}')

        ns_filetype = hFile['FileInfo'][0]['Type'][2] # 5 for ns5

        if ns_filetype == '1':
            sr = 500
        elif ns_filetype == '2':
            sr = 1000
        elif ns_filetype == '3':
            sr = 2000
        elif ns_filetype == '4':
            sr = 10000
        elif ns_filetype in ['5', '6']:
            sr = 30000
        else:
            raise ValueError(f'ERROR: {hFile.FileInfo.Type} file type not supported')

        nchan = len(hFile['Entity'])  # number of channels
        samples_per_channel = np.ceil(max_memo / (nchan * 2))
        if fi == 0:
            # get info about channels in nsx file
            chs_info = {'unit': [hFile['Entity'][i]['Units'] for i in range(nchan)],
                        'label': [hFile['Entity'][i]['Label'] for i in range(nchan)],
                        'conversion': [hFile['Entity'][i]['Scale'] for i in range(nchan)],
                        'id': [hFile['Entity'][i]['ElectrodeID'] for i in range(nchan)],
                        'pak_list': [0] * nchan,
                        'dc': [0] * nchan,
                        'macro': [hFile['Entity'][i]['Label'] for i in range(nchan)],
                        'output_name': [hFile['Entity'][i]['Label'] for i in range(nchan)]}
            micros = [i for i, x in enumerate(chs_info['unit']) if x == 'uV']
            if macro is not None and len(macro):
                chs_info['macro'] = [macro[int(np.ceil(x / 9))] if micros[i] else chs_info['macro'][i] for i, x in enumerate(range(nchan))]

            outfile_handles = [None] * nchan  # some will be empty
            _, fext = os.path.splitext(filename)
            fext = fext.lower()
            nsx_ext = fext[-1]
            ch_ext = '.NC' + nsx_ext
            if channels is None or channels == []:
                channels = hFile['FileInfo'][0]['ElectrodeList']

            remove_channels_by_label = ['(ref(.*))$']
            for ci in range(len(channels)):
                if re.search(remove_channels_by_label[0], chs_info['label'][ci]):
                    remove_chs.append(channels[ci])
            remove_chs = list(set(remove_chs))

            if remove_chs != []:
                channels = list(set(channels) - set(remove_chs))
            parsed_chs = []
            new_channel_id = []
            for i in range(nchan):
                c = hFile['FileInfo'][0]['ElectrodeList'][i]
                if c in channels:
                    ccname = c
                    if metadata != []:  # NSx file in current folder
                        repetead = [x for x in metadata['NSx'] if (x['chan_ID'] == ccname) and (x['sr'] == sr)]
                        if len(repetead):  # found channel
                            pos = [i for i, x in enumerate(repetead) if overwrite]
                            if len(pos):
                                f2delete = os.path.join(os.getcwd(), repetead[pos[0]]['output_name'] + repetead[pos[0]]['ext'])
                                print(f'Overwritting channel {metadata["NSx"][pos[0]]["chan_ID"]}. Deleting file {f2delete}')
                                if os.path.exists(f2delete):
                                    os.remove(f2delete)
                            else:
                                print(f'Skipping channel {c}, already parsed.')
                                continue  # If output_name wasn't set, the existing parsed channels won't be overwritten.
                    parsed_chs.append(c)
                    new_channel_id.append(ccname)

                    ix = chs_info['id'].index(c)
                    output_name = chs_info['macro'][ix]
                    outn_i = re.search(expr2remove, output_name)
                    if outn_i is not None and outn_i.start() > 0:
                        output_name = output_name[:outn_i.start() - 1]
                    folder_path = os.path.dirname(filename)
                    chs_info['output_name'][ix] = f'{output_name}_{ccname}'
                    nc_file_path = f"{folder_path}/{chs_info['output_name'][ix] + ch_ext}"
                    
                    outfile_handles[i] = open(nc_file_path, 'wb')

            new_files[fi]['first_sample'] = 1
        else:
            new_files[fi]['first_sample'] = new_files[fi-1]['lts'] + new_files[fi-1]['first_sample']
        if not parsed_chs:
            print('Without channels to parse.')
            return
        lts = hFile['TimeSpan']/(30000/sr)
        new_files[fi]['lts'] = lts
        N = lts
        num_segments = math.ceil(N/samples_per_channel)
        samples_per_segment = min(samples_per_channel, N)
        print(f'Data will be processed in {num_segments} segments of {samples_per_segment} samples each.')
        # min_valid_val = np.zeros((nchan, 1))
        # max_valid_val = np.zeros((nchan, 1))
        # for i in range(nchan):
        #     _, nsAnalogInfo = ns_GetAnalogInfo(hFile, i)
        #     min_valid_val[i] = nsAnalogInfo['MinVal']/nsAnalogInfo['Resolution']
        #     max_valid_val[i] = nsAnalogInfo['MaxVal']/nsAnalogInfo['Resolution']

        t_seg = time.time()
        for j in range(num_segments):
            ini = (j)*samples_per_segment
            fin = min((j+1)*samples_per_segment, N)
            _, Data = ns_GetAnalogDataBlock(hFile, list(range(0, nchan)), ini, fin-ini, 'unscale')
            for i in range(nchan):
                if outfile_handles[i]:
                    if sr == 30000:
                        pak_lost = Data[:,i] == np.iinfo(np.int16).min
                        Data[pak_lost,i] = 0
                        chs_info['pak_list'][i] = chs_info['pak_list'][i] + np.sum(pak_lost)
                    else:
                        pak_lost = Data[:,i] == np.finfo(np.float32).max
                        Data[pak_lost,i] = 0
                        chs_info['pak_list'][i] = chs_info['pak_list'][i] + np.sum(pak_lost)
                    Data_ch = Data[:,i].astype(float)
                    if sr != 30000:
                        chs_info['dc'][i] = (np.max(Data_ch) + np.min(Data_ch))/2
                        Data_ch = Data_ch - chs_info['dc'][i]
                        chs_info['conversion'][i] = np.max(np.abs(Data_ch))/np.iinfo(np.int16).max
                        Data_ch = np.round(Data_ch/chs_info['conversion'][i])
                    outfile_handles[i].write(Data_ch.astype(np.int16).tobytes())
            t_seg_mins, t_seg_secs = divmod(time.time() - t_seg, 60)
            print(f'Segment {j+1} out of {num_segments} processed in {t_seg_mins:.1f} mins and {t_seg_secs:.2f} secs.')
            t_seg = time.time()
        
        for fh in outfile_handles:
            if fh is not None:
                fh.close()
        if not metadata:
            dtype = np.dtype([('chan_ID', 'O'), ('conversion', 'O'), ('dc', 'O'), 
                              ('label', 'O'), ('bundle', 'O'), ('is_micro', 'O'), 
                              ('unit', 'O'), ('electrode_ID', 'O'), ('which_system', 'O'), 
                              ('ext', 'O'), ('lts', 'O'), ('filename', 'O'), 
                              ('sr', 'O'), ('output_name', 'O'), ('pak_list', 'O')])
            NSx = np.zeros((len(new_channel_id),), dtype=dtype)
            files_dtype = np.dtype([('name', 'O'), ('first_sample', 'O'), ('lts', 'O')])
            files = np.zeros((len(new_files),), dtype=files_dtype)
        else:
            NSx = metadata['NSx']
            files = metadata['files'].flatten()
            old_NSx_shape = NSx.shape[0]

        lts = sum([new_files[i]['lts'] for i in range(len(new_files))])
        print(f'{lts} data points written per channel')

        for ci in range(len(new_channel_id)):
            ch = new_channel_id[ci]
            elec_id = parsed_chs[ci]
            
            ix = np.where(chs_info['id']==elec_id)[0][0]
            info = {
                'chan_ID': ch,
                'conversion': chs_info['conversion'][ix],
                'dc': chs_info['dc'][ix],
                'label': chs_info['label'][ix],
                'bundle': get_bundle(chs_info['macro'][ix]),
                'is_micro': chs_info['label'][ix][0] == 'm',
                'unit': chs_info['unit'][ix],
                'electrode_ID': elec_id,
                'which_system': 'RIPPLE',
                'ext': ch_ext,
                'lts': lts,
                'filename': filenames,
                'sr': sr,
                'output_name': chs_info['output_name'][ix],
                'pak_list': chs_info['pak_list'][ix]
            }
            nsx_idx = None
            for i, nsx in enumerate(NSx):
                if nsx['chan_ID']==ch and nsx['sr']==sr:
                    # replace nsx with info
                    nsx_idx = i
                    copy_info(info, nsx)
                    break
            
            if nsx_idx is not None:
                print(f'Updated channel {ch} info.')
            else:
                if metadata: #metadata existed
                    ind = ci + old_NSx_shape
                    new_size = ind + 1  # New size must be large enough
                    dtype = np.dtype([('chan_ID', 'O'), ('conversion', 'O'), ('dc', 'O'), 
                              ('label', 'O'), ('bundle', 'O'), ('is_micro', 'O'), 
                              ('unit', 'O'), ('electrode_ID', 'O'), ('which_system', 'O'), 
                              ('ext', 'O'), ('lts', 'O'), ('filename', 'O'), 
                              ('sr', 'O'), ('output_name', 'O'), ('pak_list', 'O')])
                    new_NSx = np.empty(new_size, dtype=dtype)  # Create new array
                    new_NSx[: NSx.shape[0]] = NSx  # Copy existing data
                    NSx = new_NSx
                else:
                    ind = ci
                
                copy_info(info, NSx[ind])

        for i in range(len(new_files)):
            file_info = {
                'name': new_files[i]['name'],
                'first_sample': new_files[i]['first_sample'],
                'lts': new_files[i]['lts']
            }
            file_idx = None
            for idx, file in enumerate(files):
                if file['name'] == new_files[i]['name']:
                    file_idx = idx
                    copy_files(new_files[i], file)
                    break
            
            if file_idx is not None:
                print(f'Updated file {new_files[i]["name"]} info.')
            else:
                copy_files(new_files[i], files[i])

        freq_priority = [30000, 7500,2000, 10000, 1000, 500]

        metadata = {'NSx': NSx, 'files': files, 'freq_priority': freq_priority, 'Date_Time': Date_Time}

        scipy.io.savemat(metadata_file, metadata)

        time_taken = time.time() - t_begin
        t_mins, t_secs = divmod(time_taken, 60)
        print(f'{os.path.basename(filenames[fi])} parsed in {t_mins:.1f} minutes and {t_secs:.2f} seconds.')
        
        custompath.rm()

def copy_info(frm, to):
    to['chan_ID'] = frm['chan_ID']
    to['conversion'] = frm['conversion']
    to['dc'] = frm['dc']
    to['label'] = frm['label']
    to['bundle'] = frm['bundle']
    to['is_micro'] = frm['is_micro']
    to['unit'] = frm['unit']
    to['electrode_ID'] = frm['electrode_ID']
    to['which_system'] = frm['which_system']
    to['ext'] = frm['ext']
    to['lts'] = frm['lts']
    to['filename'] = frm['filename']
    to['sr'] = frm['sr']
    to['output_name'] = frm['output_name']
    to['pak_list'] = frm['pak_list']

def copy_files(frm, to):
    to['name'] = frm['name']
    to['first_sample'] = frm['first_sample']
    to['lts'] = frm['lts']

if __name__ == '__main__':
    
    # file_path = os.path.join(os.getcwd(), 'ns_Data', 'EMU-001_subj-MCW-FH_test_task-gaps\EMU-001_subj-MCW-FH_test_task-gaps_run-01_RIP.ns5')
    #file_path = os.path.join(os.getcwd(), 'ns_Data', 'EMU-004_subj-MCW-FH_002_task-gaps_run-01_RIP.ns5')
    #if os.path.exists(file_path):
    #    parse_ripple(file_path, overwrite=True)
    #else:
    #    print(f'File {file_path} not found.')
    file = '/mnt/data0/sEEG_DATA/MCW-FH_006/EMU/EMU-015_subj-MCW-FH_006_task-gaps/EMU-001_subj-MCW-FH_006_task-gaps_run-04_RIP.nf3'
    folder_name = os.path.dirname(file) # Folder where the orig .nf3 resides
    run_folder_name = os.path.basename(file)[:-8] # New folder name (run-xx)
    run_folder_path = os.path.join(folder_name, run_folder_name) # New folder path
    os.makedirs(run_folder_path, exist_ok=True) # Create new folder with new folder path
    
    nf3_file = os.path.join(run_folder_path, os.path.basename(file)) # Path of the nf3 copy
    if not os.path.exists(nf3_file):
        shutil.copy2(file, run_folder_path)

    parse_ripple(filenames=nf3_file, overwrite=True)

File: Python_Pipeline/codes_emu/codes_for_analysis/new_processing_pipeline/parse_ripple.m
function parse_ripple(filenames,remove_chs,macro,max_memo_GB, overwrite, channels,which_system_micro,hifreq_raw)

% This code requires the neuroshape library in the path.
% max_memo_GB is an idea of the number of GB allocated for the data to be
% stored in RAM, so it is used to compute the number of segments in which
% the data should be split for processing

if ~exist('which_system_micro','var')|| isempty(which_system_micro),  which_system_micro = 'RIP'; end 
if ~exist('hifreq_raw','var')|| isempty(hifreq_raw),  hifreq_raw = 'hifreq'; end

% [~,name] = system('hostname');
% % current_user = getenv('USER');    dir_base = sprintf('/home/%s/Documents/GitHub/codes_emu',current_user); 
% % addpath(dir_base);
% % custompath = reylab_custompath({'tasks/locations/'});

% if contains(name,'REYLAB')
%     params = MCW_location(['MCW-' which_system_micro]);
%     hours_offset = params.offset;
% end

if ~exist('hours_offset','var')|| isempty(hours_offset), hours_offset = 0; end % time difference with Coordinated Universal Time (UTC)

if ~exist('filenames','var')
   aux=dir('*.ns5');
   filenames= {aux.name};
end

expr2remove = '-\d+$';
%%
if ~isempty(mfilename)
    root_rc = [fileparts(mfilename('fullpath')) filesep '..' filesep '..'];
    if exist([root_rc filesep 'reylab_custompath.m'],'file')
        addpath(root_rc);
        custom_path = reylab_custompath('neuroshare');
    end
end

%%


if ~exist('overwrite','var') || isempty(overwrite)
    overwrite = false;
end

if ~exist('macro','var')
    macro = [];
end

if ~exist('remove_chs','var')
    remove_chs = [];
end

if ispc
    [~, systemview] = memory;
    memo_available = floor(systemview.PhysicalMemory.Available*0.80);
elseif ismac
        memo_available = 12*(1024)^3; % reduce if necessary
%     command='(top -l 1 | grep PhysMem: | awk ''{print $6}'')';
%     [status,cmdout] = system(command);
%     if status == 0
%         memo_available = floor(str2double(cmdout(1:end-3))*0.8)*(1024)^3;
%     else
%         memo_available = 12*(1024)^3; % reduce if necessary
%     end
else
    command='(free -h | awk ''/^Mem:/ {print $7}'')';
    [status,cmdout] = system(command);
    if status == 0
        memo_available = floor(str2double(cmdout(1:end-3))*0.8)*(1024)^3;
    else
        memo_available = 12*(1024)^3; % reduce if necessary
    end
end

if exist('max_memo_GB','var') && ~isempty(max_memo_GB)
    max_memo = max_memo_GB*(1024)^3;
    if max_memo > memo_available
        error('max_memo_GB > 80% of Physical Memory Available')
    end
else
    max_memo = memo_available;
end

tcum=0;

if ischar(filenames)
    filenames = {filenames};
end

% formatvector=@(v,f) sprintf(['[' repmat(['%' f ', '], 1, numel(v)-1), ['%' f ']\n']  ],v);

metadata_file = fullfile(pwd, 'NSx.mat');
if exist(metadata_file,'file')
    metadata = load(metadata_file);
else
    metadata = [];
end

for fi = 1:length(filenames)
    filename= filenames{fi};
    new_files(fi).name = filename;
    if length(filename)<3 || (~strcmpi(filename(2:3),':\') && ...
                     ~strcmpi(filename(1),'/') && ...
                     ~strcmpi(filename(2),'/') && ...
                     ~strcmpi(filename(1:2), '\\')&& ~strcmpi(filename(2:3),':/'))
        filename= [pwd filesep filename];
    end
    
    tic
     [ns_status, hFile] = ns_OpenFile(filenames{1}, 'single');  
%     [ns_status, hFile] = ns_OpenFile(filename, 'single');
%     [ns_status_nev, hFile_nev] = ns_OpenFile([filename(1:end-3) 'nev'], 'single');
%     [ns_RESULT, nsFileInfo] = ns_GetFileInfo(hFile_nev);
    
    fid = fopen(filenames{1}, 'rb');
%     fid = fopen([filename(1:end-3) 'nev'], 'rb');
    fseek(fid, 294, -1);
%     fseek(fid, 28, -1);
    Date = fread(fid, 8, 'uint16');
%     nsFileInfo.Time_Year = Date(1);
%     nsFileInfo.Time_Month = Date(2);
%     nsFileInfo.Time_Day = Date(4);
%     nsFileInfo.Time_Hour = Date(5);
%     nsFileInfo.Time_Min = Date(6);
%     nsFileInfo.Time_Sec = Date(7);
    
    tUTC = datetime([Date(1:2);Date(4:7)]','TimeZone','America/Chicago');
    [dt,dst] = tzoffset(tUTC);
    Start_Time = tUTC+dt;

%     Start_Time = datetime([Date(1:2);Date(4:7)]')-hours(hours_offset); 
%     Date_Time = sprintf('%d/%d/%d %d:%d:%d',Date(2),Date(4),Date(1),Date(5),Date(6),Date(7));
%     Date_Time = sprintf('%d/%d/%d %d:%d:%d',nsFileInfo.Time_Month,nsFileInfo.Time_Day,nsFileInfo.Time_Year,nsFileInfo.Time_Hour,nsFileInfo.Time_Min,nsFileInfo.Time_Sec);

    Rec_length_sec = hFile.TimeSpan/30000;
%     End_Time = datestr(datenum(Start_Time) + Rec_length_sec/86400, 'mm/dd/yyyy HH:MM:SS');
    End_Time = Start_Time + seconds(Rec_length_sec);

    Date_Time = [Start_Time;End_Time];
    
    if strcmp(ns_status,'ns_FILEERROR')
        error('Unable to open file: %s',filename)
    end
    
    sr=str2double(hFile.FileInfo.Label(1:strfind(hFile.FileInfo.Label,'ksamp')-2))*1000;
    electrodes_row = [];
    if sr == 7500
        for i = 1:length(hFile.Entity)
            if contains(hFile.Entity(i).Label,hifreq_raw)
                electrodes_row = horzcat(electrodes_row, i);
            end
        end
    else
        electrodes_row = 1:length(hFile.Entity);
    end
    %     switch hFile.FileInfo.Type(3)
%         case '1'
%             sr = 500;
%         case '2'
%             sr = 1000;
%         case '3'
%             sr = 2000;
%         case '4'
%             sr = 10000;            
%         case {'5' ,'6'}
%             sr = 30000;
%         otherwise
%             error('ERROR: %s file type not supported',hFile.FileInfo.Type)
%     end
    
    nchan = size(hFile.Entity,2);   % number of channels
    nchan = size(electrodes_row,2);
    samples_per_channel = ceil(max_memo/(nchan*2));
    if fi == 1
        %get info about channels in nsx file
        chs_info = struct();
        chs_info.unit = cellfun(@(x) deblank(x),{hFile.Entity(electrodes_row).Units},'UniformOutput',false);
        chs_info.label = cellfun(@(x) deblank(x),{hFile.Entity(electrodes_row).Label},'UniformOutput',false)';
        chs_info.conversion = cell2mat({hFile.Entity(electrodes_row).Scale});
        chs_info.id = cell2mat({hFile.Entity(electrodes_row).ElectrodeID});
        chs_info.pak_list = 0*chs_info.id;
        chs_info.dc= double(0*chs_info.id);
        chs_info.macro = chs_info.label;
        micros = cellfun(@(x) strcmp(x,'uV'),chs_info.unit);
        if ~isempty(macro)
            chs_info.macro(micros) = arrayfun(@(x) macro{ceil(x/9)},find(micros),'UniformOutput',false);
        end
        
        outfile_handles = cell(1,nchan); %some will be empty
        [~,~,fext] = fileparts(filename);
        fext = lower(fext(2:end));
        nsx_ext = fext(end);
        ch_ext = ['.NC' nsx_ext];
        if ~exist('channels','var') || isempty(channels)
            channels = hFile.FileInfo.ElectrodeList(electrodes_row);
        end

        remove_channels_by_label = {'(ref(.*))$'};
        for ci=1:numel(channels)
            if ~isempty(regexp(chs_info.label{ci},remove_channels_by_label{1},'match'))
                remove_chs = [remove_chs channels(ci)];
            end
        end
        remove_chs = unique(remove_chs);

        if ~isempty(remove_chs)
            channels = setdiff(channels, remove_chs);
        end
        parsed_chs = [];
        new_channel_id = [];
        k=0;
        for i = 1:length(hFile.Entity)
           if ismember(i,electrodes_row)
                k = k+1;
                c = hFile.FileInfo.ElectrodeList(i);
                if ismember(c,channels)
                    ccname = c;
                    if ~isempty(metadata) %NSx file in current folder
                        if sr == 7500
                            repetead = arrayfun(@(x) (x.chan_ID==ccname) && endsWith(x.label,hifreq_raw),metadata.NSx);
                        else 
                            repetead = arrayfun(@(x) (x.chan_ID==ccname) && (x.sr==sr),metadata.NSx);
                        end
                        if ~isempty(repetead) && sum(repetead)>0 %found channel
                            pos = find(repetead);
                            if overwrite
                                f2delete = [metadata.NSx(pos).output_name  metadata.NSx(pos).ext];
                                fprintf('Overwritting channel %d. Deleting file %s\n', metadata.NSx(pos).chan_ID, f2delete)
                                delete(f2delete)
                            else
                                fprintf('Skipping channel %d, already parsed.\n', metadata.NSx(pos).chan_ID)
                                continue %If output_name wasn't set, the existing parsed channels won't be overwritten.
                            end
                        end
                    end
                    parsed_chs(end+1) = c;
                    new_channel_id(end+1) = ccname;
                    
                    
                    ix = chs_info.id==c;
                    
                    %output_name = chs_info.label{ix};
                    output_name = chs_info.macro{ix};
                    outn_i = regexp(output_name,expr2remove);
                    if ~isempty(outn_i) && outn_i>1
                        output_name = output_name(1:outn_i-1);
                    end
                    %                 ix = find(ix,1);
                    chs_info.output_name{ix} = [output_name '_' num2str(ccname)];
                    outfile_handles{k} = fopen([chs_info.output_name{ix} ch_ext],'w');
                end
           end
        end

        new_files(fi).first_sample = 1;
    else
        new_files(fi).first_sample = new_files(fi-1).lts + new_files(fi-1).first_sample;
    end
    if isempty(parsed_chs)
        disp('Without channels to parse.')
        return
    end
   
    %total lenght adding the zeros from Timestamp
    lts = hFile.TimeSpan/(30000/sr);
    new_files(fi).lts = lts;

    N = lts;   % total data points
    num_segments = ceil(N/samples_per_channel);
    samples_per_segment = min(samples_per_channel,N);
    fprintf('Data will be processed in %d segments of %d samples each.\n',num_segments,samples_per_segment)
    
%     min_valid_val = zeros([nchan,1]);
%     max_valid_val = zeros([nchan,1]);
%     for i = 1:nchan
%         [~, nsAnalogInfo] = ns_GetAnalogInfo(hFile, i); %ns5: min -8191 max 8191 resol 0.25; nf3 min 0 max 17920 resol 1 
%         min_valid_val(i) = nsAnalogInfo.MinVal/nsAnalogInfo.Resolution;
%         max_valid_val(i) = nsAnalogInfo.MaxVal/nsAnalogInfo.Resolution;
%     end
    
    for j=1:num_segments
        ini = (j-1)*samples_per_segment+1;
        fin = min(j*samples_per_segment,N);
        tcum = tcum + toc;  % this is because openNSx has a tic at the beginning
        [~, Data] = ns_GetAnalogDataBlock(hFile, electrodes_row, ini, fin-ini+1, 'unscale'); %ns5 Data is int16, ns3 Data is single
        for i = 1:nchan
            if ~isempty(outfile_handles{i}) %channels with empty outfile_handles{i} are not selected
                if sr==30000
%                     pak_lost = Data(:,i)<min_valid_val(i);
                    pak_lost = Data(:,i) == intmin('int16');
                    Data(pak_lost,i)=0;
                    chs_info.pak_list(i) = chs_info.pak_list(i)+sum(pak_lost);  
                else
%                     pak_lost = Data(:,i) > 3e+38; % maximum value for a single-precision floating-point number 
                    pak_lost = Data(:,i) == realmax('single');
                    Data(pak_lost,i)=0;
                    chs_info.pak_list(i) = chs_info.pak_list(i)+sum(pak_lost);
                end
                Data_ch = double(Data(:,i));
                if sr~=30000
                    chs_info.dc(i)=(max(Data_ch)+min(Data_ch))/2;
                    Data_ch=Data_ch- chs_info.dc(i);
                    chs_info.conversion(i) = max(abs(Data_ch))/double(intmax('int16'));
                    Data_ch = round(Data_ch/chs_info.conversion(i));
                end
                fwrite(outfile_handles{i},int16(Data_ch),'int16');
            end
        end
        fprintf('Segment %d out of %d processed. ',j,num_segments)
    end
    
    tcum = tcum + toc;
    fprintf('Total time spent in parsing the data was %s secs.\n',num2str(tcum, '%0.1f')); 
end
fclose('all');



if isempty(metadata)
    files = [];
    NSx = [];
else
    NSx = metadata.NSx;
    files = metadata.files;
end
lts = sum([new_files(:).lts]);
fprintf('%d data points written per channel\n',lts)


for ci = 1:length(new_channel_id)
    ch = new_channel_id(ci);
    elec_id = parsed_chs(ci);
    if sr == 7500
        repetead = arrayfun(@(x) (x.chan_ID==ch) && endsWith(x.label,hifreq_raw),NSx);
    else
        repetead = arrayfun(@(x) (x.chan_ID==ch) && (x.sr==sr),NSx);
    end
    if isempty(repetead) || sum(repetead)==0
        pos = length(NSx)+1;
    else
        pos = find(repetead);
    end
        ix = chs_info.id==elec_id;
%         ix = find(ix,1);
        NSx(pos).chan_ID = ch;
        NSx(pos).conversion = chs_info.conversion(ix);
        NSx(pos).dc = chs_info.dc(ix);
        NSx(pos).label = chs_info.label{ix};
        NSx(pos).bundle = get_bundle(chs_info.macro{ix});
%         NSx(pos).is_micro = NSx(pos).label(1)=='m';
        NSx(pos).unit = chs_info.unit{ix};
        NSx(pos).electrode_ID = elec_id;
        NSx(pos).which_system = 'RIPPLE';
        NSx(pos).ext = ch_ext;
        NSx(pos).lts = lts;
        NSx(pos).filename = filenames;
        NSx(pos).sr = sr;
        NSx(pos).output_name = chs_info.output_name{ix};
        NSx(pos).pak_list =chs_info.pak_list(ix);
        NSx(pos).is_micro =strcmp(NSx(pos).unit,'uV') && NSx(pos).sr==30000;
end

for i = 1:length(new_files)
    repetead = arrayfun(@(x) strcmp(x.name,new_files(i).name),files);
    if isempty(repetead) || sum(repetead)==0
        pos = length(files)+1;
    else
        pos = find(repetead);
    end
    files(pos).name = new_files(i).name;
    files(pos).first_sample = new_files(i).first_sample;
    files(pos).lts = new_files(i).lts;
end
freq_priority=[30000, 2000, 7500, 10000, 1000, 500];
save(metadata_file, 'NSx','files','freq_priority','Date_Time')
% % custom_path.rm()
end

function bundle=get_bundle(label)
    tt = regexp(label,'\d+','once');
    if ~isempty(tt)
        bundle = label(1:tt-1);
    else
        bundle = label;
    end    
%     macro = regexp(label,'^\w*(?=(\d* raw$))','match');
%     if ~isempty(macro)
%         macro = macro{1};
%     else
%         macro = label;
%     end
end
